File: algorithms/python/全排列/全排列.py
# Source : https://leetcode.cn/problems/permutations/description/
# Author : Yanan Wang
# Date   : 2023-08-31

# 本题先画出一个决策树，然后根据该树写递归
# 用全局变量ans保存结果，用path记录下递归过程中的数字，用used进行剪枝
# 用Python写本题时，注意递归出口那里，python传参用的是引用传参(浅拷贝)，这里应该用深拷贝

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans, path, used = [], [], [False for _ in range(len(nums))]
        def dfs(nums):
            if len(path) == len(nums):
                ans.append(path[:])
                return
            for i in range(len(nums)):
                if not used[i]:
                    path.append(nums[i])
                    used[i] = True
                    dfs(nums)
                    path.pop()
                    used[i] = False

        dfs(nums)
        return ans

File: algorithms/cpp/全排列/全排列.cpp
// Source : https://leetcode.cn/problems/permutations/description/ 
// Author : Yanan Wang
// Date   : 2023-11-29

class Solution {
public:
    int n;
    vector<bool> vis;
    vector<int> path;
    vector<vector<int>> ans; 
    void dfs(vector<int> &nums) {
        if (path.size() == n) {
            ans.push_back(path);
            return;
        }
        for (int i = 0; i < n; ++i) {
            if (!vis[i]) {
                path.push_back(nums[i]);
                vis[i] = true;
                dfs(nums);
                path.pop_back();
                vis[i] = false; 
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        n = nums.size();
        vis.resize(n);
        dfs(nums);
        return ans;
    }
};